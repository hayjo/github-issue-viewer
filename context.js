import { createContext } from 'react';

const RepoContext = createContext();

export { RepoContext };
