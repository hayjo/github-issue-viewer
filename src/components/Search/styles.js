// @flow

import { StyleSheet } from 'react-native';
import type {
  ViewStyleProp,
  TextStyleProp,
} from 'react-native/Libraries/StyleSheet/StyleSheet';

import { COLORS } from '../../constants';

const viewStyles: { [string]: ViewStyleProp } = StyleSheet.create({
  container: {
    height: '100%',
    width: '100%',
    flex: 1,
  },
  searchBarContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginVertical: 12,
    marginHorizontal: 12,
  },
  searchBar: {
    flexDirection: 'row',
    flex: 1,
    marginHorizontal: 4,
    marginVertical: 2,
    backgroundColor: COLORS.SEARCH_BACKGROUND,
    borderRadius: 8,
    overflow: 'hidden',
    paddingRight: 4,
  },
  mainContainer: {
    flex: 1,
  },
  cardContainer: {
    paddingHorizontal: 16,
  },
  cardContent: {
    padding: 2,
    flex: 1,
    overflow: 'hidden',
  },
  checkIconContainer: {
    marginVertical: 5,
    marginLeft: 4,
    alignSelf: 'flex-start',
  },
});

const textStyles: { [string]: TextStyleProp } = StyleSheet.create({
  cardTitle: {
    color: COLORS.TITLE,
    fontSize: 16,
  },
  resultCount: {
    marginBottom: 12,
    fontSize: 16,
    color: COLORS.SUBTITLE,
  },
  searchIcon: {
    alignSelf: 'center',
    marginHorizontal: 8,
    color: COLORS.ICON,
  },
  repoIcon: {
    marginLeft: 2,
    marginRight: 8,
    marginVertical: 5,
    color: COLORS.ICON,
  },
});

export { viewStyles, textStyles };
