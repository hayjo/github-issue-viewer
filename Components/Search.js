import React, { useContext, useState, useEffect, useRef } from 'react';
import {
  SafeAreaView,
  FlatList,
  ActivityIndicator,
  View,
  Text,
  TextInput,
  Button,
  StyleSheet,
} from 'react-native';
import type { Element } from 'react';
import Icon from 'react-native-vector-icons/Octicons';
import debounce from 'lodash.debounce';
import Card from './Card';
import SnackBar from './SnackBar';
import { RepoContext } from '../context';
import { searchRepoByQuery } from '../api';
import { COLORS, MESSAGE, SIZE, LIMIT } from '../constants';

Icon.loadFont();

const Search: () => Element = ({ onCancel }) => {
  const {
    selectedRepoList,
    onSelectRepo,
    hasExceedLimit,
    onNoticeClick,
    notifyError,
  } = useContext(RepoContext);
  const [searchQuery, setSearchQuery] = useState('');
  const [nextPage, setNextPage] = useState(1);
  const [resultCount, setResultCount] = useState(0);
  const [repoList, setRepoList] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const ref = useRef(null);
  const debouncingTime = nextPage === 1 ? 0 : LIMIT.SEARCH_DEBOUNCING_TIME;

  const onChangeSearch = query => setSearchQuery(query);

  const searchRepositories = debounce(async (query, pageNumber) => {
    if (!query) {
      return;
    }

    if (isLoading) {
      return;
    }

    setIsLoading(true);

    try {
      const searchResult = await searchRepoByQuery(query, pageNumber);
      const { totalCount, items } = searchResult;

      setRepoList(prevList => [...prevList, ...items]);
      setResultCount(totalCount);

      if (items.length) {
        setNextPage(pageNumber + 1);
      }
    } catch (err) {
      notifyError(err);
    } finally {
      setIsLoading(false);
    }
  }, debouncingTime);

  const renderItem = ({ item }) => {
    const { id, name, owner, description } = item;
    const { id: ownerId, login: ownerName } = owner;
    const fullName = `${ownerName}/${name}`;

    return (
      <Card
        key={id}
        onPress={() => onSelectRepo({ id, name, ownerId, ownerName })}>
        <Icon style={styles.repoIcon} name="repo" size={SIZE.ICON} />
        <View style={styles.cardContent}>
          <Text style={styles.cardTitle}>{fullName}</Text>
          <Text numberOfLines={2}>{description}</Text>
        </View>
        {selectedRepoList.find(repo => repo.id === id) ? (
          <View style={styles.checkIcon}>
            <Icon name="check" size={SIZE.ICON} color={COLORS.SELECTED} />
          </View>
        ) : null}
      </Card>
    );
  };

  const listHeader = (
    <Text style={styles.resultCount}>
      {resultCount.toLocaleString()} repository results
    </Text>
  );

  const listFooter = isLoading ? <ActivityIndicator /> : null;

  useEffect(() => {
    if (ref?.current) {
      ref.current.focus();
    }
  }, []);

  useEffect(() => {
    setRepoList([]);
    setNextPage(1);
  }, [searchQuery]);

  return (
    <SafeAreaView style={styles.container}>
      <View style={styles.searchBarContainer}>
        <View style={styles.searchBar}>
          <Icon name="search" style={styles.searchIcon} size={16} />
          <TextInput
            ref={ref}
            placeholder="Search Repository"
            onChangeText={onChangeSearch}
            value={searchQuery}
            onSubmitEditing={() => searchRepositories(searchQuery, nextPage)}
          />
        </View>
        <Button title="Cancel" onPress={onCancel} />
      </View>
      <View style={styles.mainContainer}>
        <FlatList
          data={repoList}
          style={styles.cardContainer}
          renderItem={renderItem}
          keyExtractor={({ id }) => id}
          ListHeaderComponent={listHeader}
          ListFooterComponent={listFooter}
          onEndReached={() => searchRepositories(searchQuery, nextPage)}
        />
      </View>
      {hasExceedLimit ? (
        <SnackBar
          onPress={() => onNoticeClick()}
          text={MESSAGE.MAX_REPO_COUNT}
        />
      ) : null}
    </SafeAreaView>
  );
};

const styles = StyleSheet.create({
  container: {
    height: '100%',
    width: '100%',
    flex: 1,
  },
  searchBarContainer: {
    flexDirection: 'row',
    justifyContent: 'space-between',
    marginVertical: 12,
    marginHorizontal: 12,
  },
  searchBar: {
    flexDirection: 'row',
    flex: 1,
    marginHorizontal: 4,
    marginVertical: 2,
    backgroundColor: COLORS.SEARCH_BACKGROUND,
    borderRadius: 8,
    overflow: 'hidden',
    paddingRight: 4,
  },
  searchIcon: {
    alignSelf: 'center',
    marginHorizontal: 8,
    color: COLORS.ICON,
  },
  mainContainer: {
    flex: 1,
  },
  cardContainer: {
    paddingHorizontal: 16,
  },
  resultCount: {
    marginBottom: 12,
    fontSize: 16,
    color: COLORS.SUBTITLE,
  },
  repoIcon: {
    marginLeft: 2,
    marginRight: 8,
    marginVertical: 5,
    color: COLORS.ICON,
  },
  cardContent: {
    padding: 2,
    flex: 1,
    overflow: 'hidden',
  },
  cardTitle: {
    color: COLORS.TITLE,
    fontSize: 16,
  },
  checkIcon: {
    marginVertical: 5,
    marginLeft: 4,
    alignSelf: 'flex-start',
  },
});

export default Search;
