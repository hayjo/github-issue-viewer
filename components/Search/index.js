// @flow

import React, { useContext, useState, useEffect, useRef } from 'react';
import {
  SafeAreaView,
  FlatList,
  ActivityIndicator,
  View,
  Text,
  TextInput,
  Button,
} from 'react-native';
import type { Node } from 'react';
import Icon from 'react-native-vector-icons/Octicons';
import debounce from 'lodash.debounce';

import Card from '../common/Card';
import SnackBar from '../common/SnackBar';

import { RepoContext } from '../../context';
import { searchRepoByQuery } from '../../api';
import { COLORS, MESSAGE, SIZE, LIMIT } from '../../constants';

import { viewStyles, textStyles } from './styles';

Icon.loadFont();

type Props = {
  onCancel: () => void,
};

const Search: (props: Props) => Node = ({ onCancel }) => {
  const {
    selectedRepoList,
    onSelectRepo,
    hasExceedLimit,
    onNoticeClick,
    notifyError,
  } = useContext(RepoContext);
  const [searchQuery, setSearchQuery] = useState('');
  const [nextPage, setNextPage] = useState(1);
  const [resultCount, setResultCount] = useState(0);
  const [repoList, setRepoList] = useState([]);
  const [isLoading, setIsLoading] = useState(false);
  const ref = useRef(null);
  const debouncingTime = nextPage === 1 ? 0 : LIMIT.SEARCH_DEBOUNCING_TIME;

  const onChangeSearch = query => setSearchQuery(query);

  const searchRepositories = debounce(async (query, pageNumber) => {
    if (!query) {
      return;
    }

    if (isLoading) {
      return;
    }

    setIsLoading(true);

    try {
      const searchResult = await searchRepoByQuery(query, pageNumber);
      const { totalCount, items } = searchResult;

      setRepoList(prevList => [...prevList, ...items]);
      setResultCount(totalCount);

      if (items.length) {
        setNextPage(pageNumber + 1);
      }
    } catch (err) {
      notifyError(err);
    } finally {
      setIsLoading(false);
    }
  }, debouncingTime);

  const renderItem = ({ item }) => {
    const { id, name, owner, description } = item;
    const { id: ownerId, login: ownerName } = owner;
    const fullName = `${ownerName}/${name}`;

    return (
      <Card
        key={id}
        onPress={() => onSelectRepo({ id, name, ownerId, ownerName })}>
        <Icon style={textStyles.repoIcon} name="repo" size={SIZE.ICON} />
        <View style={viewStyles.cardContent}>
          <Text style={textStyles.cardTitle}>{fullName}</Text>
          <Text numberOfLines={2}>{description}</Text>
        </View>
        {selectedRepoList.find(repo => repo.id === id) ? (
          <View style={viewStyles.checkIconContainer}>
            <Icon name="check" size={SIZE.ICON} color={COLORS.SELECTED} />
          </View>
        ) : null}
      </Card>
    );
  };

  const listHeader = (
    <Text style={textStyles.resultCount}>
      {resultCount.toLocaleString()} repository results
    </Text>
  );

  const listFooter = isLoading ? <ActivityIndicator /> : null;

  useEffect(() => {
    if (ref?.current) {
      ref.current.focus();
    }
  }, []);

  useEffect(() => {
    setRepoList([]);
    setNextPage(1);
  }, [searchQuery]);

  return (
    <SafeAreaView style={viewStyles.container}>
      <View style={viewStyles.searchBarContainer}>
        <View style={viewStyles.searchBar}>
          <Icon name="search" style={textStyles.searchIcon} size={16} />
          <TextInput
            ref={ref}
            placeholder="Search Repository"
            onChangeText={onChangeSearch}
            value={searchQuery}
            onSubmitEditing={() => searchRepositories(searchQuery, nextPage)}
          />
        </View>
        <Button title="Cancel" onPress={onCancel} />
      </View>
      <View style={viewStyles.mainContainer}>
        <FlatList
          data={repoList}
          style={viewStyles.cardContainer}
          renderItem={renderItem}
          keyExtractor={({ id }) => String(id)}
          ListHeaderComponent={listHeader}
          ListFooterComponent={listFooter}
          onEndReached={() => searchRepositories(searchQuery, nextPage)}
        />
      </View>
      {hasExceedLimit ? (
        <SnackBar
          onPress={() => onNoticeClick()}
          text={MESSAGE.MAX_REPO_COUNT}
        />
      ) : null}
    </SafeAreaView>
  );
};

export default Search;
