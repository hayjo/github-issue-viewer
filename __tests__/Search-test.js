import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import Search from '../components/Search';
import { RepoContext } from '../context';

const initialData = {
  selectedRepoList: [],
  hasExceedLimit: false,
  onSelectRepo: jest.fn(),
  onNoticeClick: jest.fn(),
  notifyError: jest.fn(),
};

describe('Search component test', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <RepoContext.Provider value={initialData}>
          <Search onCancel={jest.fn()} />
        </RepoContext.Provider>,
      )
      .toJSON();

    expect(tree).toMatchSnapshot();
  });
});
