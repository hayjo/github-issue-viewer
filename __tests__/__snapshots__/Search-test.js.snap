// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component test renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
  style={
    Object {
      "flex": 1,
      "height": "100%",
      "width": "100%",
    }
  }
>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginHorizontal": 12,
        "marginVertical": 12,
      }
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#1b1f241a",
          "borderRadius": 8,
          "flex": 1,
          "flexDirection": "row",
          "marginHorizontal": 4,
          "marginVertical": 2,
          "overflow": "hidden",
          "paddingRight": 4,
        }
      }
    >
      <Text
        allowFontScaling={false}
        style={
          Array [
            Object {
              "color": undefined,
              "fontSize": 16,
            },
            Object {
              "alignSelf": "center",
              "color": "#6a737d",
              "marginHorizontal": 8,
            },
            Object {
              "fontFamily": "Octicons",
              "fontStyle": "normal",
              "fontWeight": "normal",
            },
            Object {},
          ]
        }
      >
        ï†—
      </Text>
      <TextInput
        onChangeText={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search Repository"
        value=""
      />
    </View>
    <View
      accessibilityRole="button"
      accessible={true}
      collapsable={false}
      focusable={true}
      nativeID="animatedComponent"
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Array [
            Object {},
          ]
        }
      >
        <Text
          style={
            Array [
              Object {
                "color": "#007AFF",
                "fontSize": 18,
                "margin": 8,
                "textAlign": "center",
              },
            ]
          }
        >
          Cancel
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RCTScrollView
      ListFooterComponent={null}
      ListHeaderComponent={
        <Text
          style={
            Object {
              "color": "#24292f",
              "fontSize": 16,
              "marginBottom": 12,
            }
          }
        >
          0
           repository results
        </Text>
      }
      data={Array []}
      getItem={[Function]}
      getItemCount={[Function]}
      keyExtractor={[Function]}
      onContentSizeChange={[Function]}
      onEndReached={[Function]}
      onLayout={[Function]}
      onMomentumScrollBegin={[Function]}
      onMomentumScrollEnd={[Function]}
      onScroll={[Function]}
      onScrollBeginDrag={[Function]}
      onScrollEndDrag={[Function]}
      removeClippedSubviews={false}
      renderItem={[Function]}
      scrollEventThrottle={50}
      stickyHeaderIndices={Array []}
      style={
        Object {
          "paddingHorizontal": 16,
        }
      }
      viewabilityConfigCallbackPairs={Array []}
    >
      <View>
        <View
          onLayout={[Function]}
        >
          <Text
            style={
              Object {
                "color": "#24292f",
                "fontSize": 16,
                "marginBottom": 12,
              }
            }
          >
            0
             repository results
          </Text>
        </View>
      </View>
    </RCTScrollView>
  </View>
</RCTSafeAreaView>
`;
